File: euler_products/extractor.hs
{-# LANGUAGE BangPatterns #-}
-- Term-by-term extraction of B_p(x) = NUM(x)/DEN(x) into a product of
-- (1 - p^l x^u)^gamma factors, matching Euler's zeta identities.
--
-- p-polynomials are sparse maps: power-of-p -> integer coefficient.
-- Series in x are plain lists of p-polynomials, index = power of x.

import qualified Data.Map.Strict as M
import Data.List (foldl')
import System.CPUTime
import Text.Printf (printf)

type PPoly  = M.Map Int Integer
type Series = [PPoly]                    -- length N+1, index k = coeff of x^k

pZero, pOne :: PPoly
pZero = M.empty
pOne  = M.singleton 0 1

pAdd :: PPoly -> PPoly -> PPoly
pAdd a b = M.filter (/= 0) (M.unionWith (+) a b)

pMul :: PPoly -> PPoly -> PPoly
pMul a b = M.filter (/= 0) $ M.fromListWith (+)
  [ (i + j, ca * cb) | (i, ca) <- M.toList a, (j, cb) <- M.toList b ]

seriesMulTrunc :: Int -> Series -> Series -> Series
seriesMulTrunc n a b =
  [ foldl' pAdd pZero [ pMul (a !! i) (b !! (k - i)) | i <- [0 .. k] ] | k <- [0 .. n] ]

-- Invert a series with constant term exactly 1 (over the ring Z[p]).
-- Fails (returns Nothing) if the constant term isn't the unit 1.
seriesInvert :: Int -> Series -> Maybe Series
seriesInvert n a
  | a !! 0 /= pOne = Nothing
  | otherwise      = Just (go [pOne])
  where
    go acc
      | length acc > n = reverse (take (n+1) (reverse acc))
      | otherwise =
          let m = length acc
              s = foldl' pAdd pZero
                    [ pMul (a !! i) (acc !! (m - i)) | i <- [1 .. m] ]
              next = M.map negate s
          in go (acc ++ [next])

-- exact integer binomial coefficient C(g,k) for any integer g (can be negative)
binom :: Integer -> Int -> Integer
binom g k = product [ g - fromIntegral i | i <- [0 .. k-1] ] `div` product [1 .. toInteger k]

-- series for (1 - p^l x^u)^c, truncated to order n
factorSeries :: Int -> Int -> Int -> Integer -> Series
factorSeries n u l c = [ term k | k <- [0 .. n] ]
  where
    term k
      | k `mod` u /= 0 = pZero
      | otherwise =
          let j = k `div` u
          in if j == 0 then pOne
             else let co = binom c j * (if odd j then -1 else 1)
                  in if co == 0 then pZero else M.singleton (l * j) co

-- At each order u, the residual coefficient c(p) = sum_l c_l p^l is a plain
-- integer polynomial: extract ONE factor (1-p^l x^u)^{-c_l} per nonzero term,
-- all at this same u, before moving to u+1.
extract :: Int -> Series -> [(Int,Int,Integer)]
extract n res0 = go 1 [] res0
  where
    go u acc res
      | u > n = reverse acc
      | otherwise =
          let terms = M.toList (res !! u)
          in if null terms
             then go (u+1) acc res
             else let (acc', res') = foldl' step (acc, res) terms
                      step (a, r) (l, c0) =
                        let gamma  = negate c0
                            factor = factorSeries n u l c0
                        in ((u,l,gamma):a, seriesMulTrunc n r factor)
                  in go (u+1) acc' res'

mkSeries :: Int -> [(Int, PPoly)] -> Series
mkSeries n assoc = [ M.findWithDefault pZero k m | k <- [0 .. n] ]
  where m = M.fromList assoc

poly :: [(Int,Integer)] -> PPoly
poly = M.filter (/=0) . M.fromList

bitLen :: Integer -> Int
bitLen x = length (show (abs x))   -- decimal digit count, good enough as a size proxy

runCase :: Bool -> String -> Int -> Series -> Series -> IO ()
runCase verbose name n num den = do
  putStrLn $ "=== " ++ name ++ " (order N=" ++ show n ++ ") ==="
  case seriesInvert n den of
    Nothing -> putStrLn "  DEN(0) != 1 in Z[p]: cannot invert termwise in the polynomial ring. STOP."
    Just denInv -> do
      let r = seriesMulTrunc n num denInv
          factors = extract n r
      if verbose
        then mapM_ (\(u,l,g) -> printf "  u=%2d  l=%3d  gamma=%4d   -> (1 - p^%d x^%d)^%d\n" u l g l u g) factors
        else do
          printf "  factors extracted: %d\n" (length factors)
          let maxDigits = maximum (0 : [ bitLen g | (_,_,g) <- factors ])
          printf "  max |gamma| decimal digits: %d\n" maxDigits
          case factors of
            [] -> return ()
            _  -> let (u,l,g) = last factors
                  in printf "  last: u=%d l=%d gamma=%d\n" u l g
  putStrLn ""

main :: IO ()
main = do
  let n = 60   -- benchmark order
      verbose = False
      num1 = mkSeries n [ (0, pOne), (1, poly [(0,1),(1,-2)]) ]
      den1 = mkSeries n [ (0, pOne), (1, poly [(2,-1)]) ]
      num2 = mkSeries n [ (0, poly [(2,1),(3,-2),(4,1)]), (1, poly [(0,-1),(1,2)]) ]
      den2 = mkSeries n [ (0, poly [(2,1),(3,-2),(4,1)]), (1, poly [(0,-1),(1,2),(2,-1)]) ]
  t0 <- getCPUTime
  runCase verbose "phi^2(n)      [Mathar 3.78/3.79 test]" n num1 den1
  runCase verbose "1/phi(n)^2    [our target]"             n num2 den2
  t1 <- getCPUTime
  printf "CPU time: %.4f s\n" (fromIntegral (t1 - t0) / (10^12) :: Double)

File: euler_products/extractor.cpp
// Term-by-term extraction of B_p(x) = NUM(x)/DEN(x) into a product of
// (1 - p^l x^u)^gamma factors. Same algorithm as extractor.hs.
//
// p-polynomial: sparse map power-of-p -> mpz_class coefficient.
// Series in x: vector of p-polynomials, index = power of x.

#include <gmpxx.h>
#include <map>
#include <vector>
#include <iostream>
#include <chrono>
#include <optional>

using PPoly  = std::map<int, mpz_class>;
using Series = std::vector<PPoly>;

static void trim(PPoly &p) {
    for (auto it = p.begin(); it != p.end(); ) {
        if (it->second == 0) it = p.erase(it); else ++it;
    }
}

PPoly pAdd(const PPoly &a, const PPoly &b) {
    PPoly r = a;
    for (auto &kv : b) r[kv.first] += kv.second;
    trim(r);
    return r;
}

PPoly pMul(const PPoly &a, const PPoly &b) {
    PPoly r;
    for (auto &ka : a)
        for (auto &kb : b)
            r[ka.first + kb.first] += ka.second * kb.second;
    trim(r);
    return r;
}

PPoly pOne() { return PPoly{{0, mpz_class(1)}}; }

Series seriesMulTrunc(int n, const Series &a, const Series &b) {
    Series r(n + 1);
    for (int k = 0; k <= n; k++) {
        PPoly acc;
        for (int i = 0; i <= k; i++) acc = pAdd(acc, pMul(a[i], b[k - i]));
        r[k] = acc;
    }
    return r;
}

// Invert a series with constant term exactly 1 (ring Z[p]); nullopt if not.
std::optional<Series> seriesInvert(int n, const Series &a) {
    if (!(a[0].size() == 1 && a[0].count(0) && a[0].at(0) == 1)) return std::nullopt;
    Series inv(n + 1);
    inv[0] = pOne();
    for (int m = 1; m <= n; m++) {
        PPoly s;
        for (int i = 1; i <= m; i++) s = pAdd(s, pMul(a[i], inv[m - i]));
        for (auto &kv : s) kv.second = -kv.second;
        trim(s);
        inv[m] = s;
    }
    return inv;
}

// exact integer binomial coefficient C(g,k) for any integer g (possibly negative)
mpz_class binom(const mpz_class &g, int k) {
    mpz_class num = 1;
    for (int i = 0; i < k; i++) num *= (g - i);
    mpz_class fact = 1;
    for (int i = 1; i <= k; i++) fact *= i;
    return num / fact; // exact
}

// series for (1 - p^l x^u)^c, truncated to order n
Series factorSeries(int n, int u, int l, const mpz_class &c) {
    Series f(n + 1);
    for (int k = 0; k <= n; k++) {
        if (k % u != 0) continue;
        int j = k / u;
        if (j == 0) { f[k] = pOne(); continue; }
        mpz_class co = binom(c, j) * (j % 2 ? -1 : 1);
        if (co != 0) f[k][l * j] = co;
    }
    return f;
}

struct Factor { int u, l; mpz_class gamma; };

std::vector<Factor> extract(int n, Series res) {
    std::vector<Factor> out;
    for (int u = 1; u <= n; u++) {
        // snapshot terms at this order before mutating res via factor division
        std::vector<std::pair<int, mpz_class>> terms(res[u].begin(), res[u].end());
        for (auto &t : terms) {
            int l = t.first;
            mpz_class c0 = t.second;
            if (c0 == 0) continue;
            mpz_class gamma = -c0;
            Series factor = factorSeries(n, u, l, c0);
            res = seriesMulTrunc(n, res, factor);
            out.push_back({u, l, gamma});
        }
    }
    return out;
}

Series mkSeries(int n, const std::vector<std::pair<int, PPoly>> &assoc) {
    Series s(n + 1);
    for (auto &kv : assoc) s[kv.first] = kv.second;
    return s;
}

PPoly poly(const std::vector<std::pair<int, long long>> &terms) {
    PPoly p;
    for (auto &t : terms) if (t.second != 0) p[t.first] = mpz_class((long)t.second);
    return p;
}

int digitCount(const mpz_class &g) {
    mpz_class a = g < 0 ? -g : g;
    return a.get_str().size();
}

void runCase(bool verbose, const std::string &name, int n, const Series &num, const Series &den) {
    std::cout << "=== " << name << " (order N=" << n << ") ===\n";
    auto denInv = seriesInvert(n, den);
    if (!denInv) {
        std::cout << "  DEN(0) != 1 in Z[p]: cannot invert termwise in the polynomial ring. STOP.\n\n";
        return;
    }
    Series r = seriesMulTrunc(n, num, *denInv);
    auto factors = extract(n, r);
    if (verbose) {
        for (auto &f : factors)
            std::cout << "  u=" << f.u << "  l=" << f.l << "  gamma=" << f.gamma
                       << "   -> (1 - p^" << f.l << " x^" << f.u << ")^" << f.gamma << "\n";
    } else {
        std::cout << "  factors extracted: " << factors.size() << "\n";
        int maxDigits = 0;
        for (auto &f : factors) maxDigits = std::max(maxDigits, digitCount(f.gamma));
        std::cout << "  max |gamma| decimal digits: " << maxDigits << "\n";
        if (!factors.empty()) {
            auto &f = factors.back();
            std::cout << "  last: u=" << f.u << " l=" << f.l << " gamma=" << f.gamma << "\n";
        }
    }
    std::cout << "\n";
}

int main() {
    int n = 60;
    bool verbose = false;

    Series num1 = mkSeries(n, { {0, pOne()}, {1, poly({{0,1},{1,-2}})} });
    Series den1 = mkSeries(n, { {0, pOne()}, {1, poly({{2,-1}})} });

    Series num2 = mkSeries(n, { {0, poly({{2,1},{3,-2},{4,1}})}, {1, poly({{0,-1},{1,2}})} });
    Series den2 = mkSeries(n, { {0, poly({{2,1},{3,-2},{4,1}})}, {1, poly({{0,-1},{1,2},{2,-1}})} });

    auto t0 = std::chrono::high_resolution_clock::now();
    runCase(verbose, "phi^2(n)      [Mathar 3.78/3.79 test]", n, num1, den1);
    runCase(verbose, "1/phi(n)^2    [our target]",             n, num2, den2);
    auto t1 = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> dt = t1 - t0;
    std::cout << "CPU time: " << dt.count() << " s\n";
    return 0;
}
